core/controllers/reader.py

```diff
@@ -122,31 +123,67 @@ def classify(exp_id, state, answer, params):
-        fs = fs_domain.AbstractFileSystem(
-            fs_domain.ExplorationFileSystem(exp_id))
-        input_type = interaction_instance.answer_type
-        ored_truth_value = 0.0
-        best_rule_spec = None
         for rule_spec in answer_group.rule_specs:
-            evaluated_truth_value = rule_domain.evaluate_rule(
-                rule_spec, input_type, params, normalized_answer, fs)
-            if evaluated_truth_value > ored_truth_value:
-                ored_truth_value = evaluated_truth_value
-                best_rule_spec = rule_spec
-        if ored_truth_value > best_matched_truth_value:
-            best_matched_truth_value = ored_truth_value
-            best_matched_rule_spec = best_rule_spec
-            best_matched_answer_group = answer_group
-            best_matched_answer_group_index = i
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > best_matched_truth_value:
+                    best_matched_truth_value = evaluated_truth_value
+                    best_matched_rule_spec = rule_spec
+                    best_matched_answer_group = answer_group
+                    best_matched_answer_group_index = i
+
+    # Find the maximum soft rule that matches. This is
+    # done by ORing (maximizing) all truth values of all rules over all answer
+    # groups. The group with the highest truth value is considered the best
+    # match.
+    if best_matched_answer_group is None:
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    evaluated_truth_value = rule_domain.evaluate_rule(
+                        rule_spec, input_type, params, normalized_answer, fs)
+                    if evaluated_truth_value > best_matched_truth_value:
+                        best_matched_truth_value = evaluated_truth_value
+                        best_matched_rule_spec = rule_spec
+                        best_matched_answer_group = answer_group
+                        best_matched_answer_group_index = i
+
+    # Run the classifier
+    if best_matched_answer_group is None:
+        sc = classifier_services.StringClassifier()
+        training_examples = [[doc, []] for doc in
+            state.interaction.confirmed_unclassified_answers]
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    training_examples.extend([[doc, [str(i)]] for doc in
+                        rule_spec.inputs['training_data']])
+        if len(training_examples) > 0:
```

Comment:
Ah, good point. I misread the initialization line of training_examples. Yes, the idea is right then. Regarding whether to do the check here or not, I think it's fine here. It makes sense that if there is no data to train the classifier on, the classifier should not be used. However, it would be worth ensuring the classifier has a test for passing nothing to it, if there isn't one already.